WordPress/Classes/Utility/Blogging Reminders/BloggingRemindersScheduler.swift

```diff
@@ -277,16 +277,16 @@ class BloggingRemindersScheduler {
         }
 
         pushNotificationAuthorizer.requestAuthorization { [weak self] allowed in
-            guard let self = self else {
-                return
-            }
-
-            guard allowed else {
-                completion(.failure(Error.needsPermissionForPushNotifications))
-                return
+            DispatchQueue.main.async {
+                guard let self = self else {
+                    return
+                }
+                guard allowed else {
+                    completion(.failure(Error.needsPermissionForPushNotifications))
+                    return
+                }
+                self.pushAuthorizationReceived(blog: blog, schedule: schedule, time: time, completion: completion)
```

Comment:
Are we sure the `blog` is _always_ bound to the main context? I can't see it in the context of this function.

Reply:
In most cases, yes.

- `JetpackNotificationMigrationService` - yes. `notificationSettingsService?.getAllSettings` uses Alamofire, which calls the callbacks on the main queue.
- `BloggingRemindersFlowSettingsViewController` - yes. One call comes from `notifyMeButtonTapped`. Another is called from `BloggingRemindersPushPromptViewController`, which has an explicit dispatch async to main.
- The call stack from `remove(blog:)` all happen either from alert confirmations or from Alamofire callbacks – both main.

There is actually one place that I initially missed in `BlogService.syncBlogsForAccount` where it is called on a different queue:

```swift
[self.coreDataStack performAndSaveUsingBlock:^(NSManagedObjectContext *context) {
    [self mergeBlogs:blogs withAccountID:account.objectID inContext:context];
} completion:success onQueue:dispatch_get_main_queue()];

// ...

- (void)mergeBlogs:(NSArray<RemoteBlog *> *)blogs withAccountID:(NSManagedObjectID *)accountID inContext:(NSManagedObjectContext *)context
{
    if ([toDelete count] > 0) {
        for (Blog *blog in account.blogs) {
            if ([toDelete containsObject:blog.dotComID]) {
                [self unscheduleBloggingRemindersFor:blog]; // <--- here
                // Consider switching this to a call to removeBlog in the future
                // to consolidate behaviour @frosty
                [context deleteObject:blog];
            }
        }
    }
}
```

So, this PR addresses the issue on the reminders screen but doesn't address it in the `mergeBlogs`. When you check authorization, the callbacks are called on an arbitrary queue which is **not** a MOC-managed queue. So `mergeBlogs` has the same issues as the reminders.

<img width="781" alt="Screenshot 2023-08-03 at 9 30 48 AM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/3f856302-c7a2-42db-91fc-c515b402e471">

I want to get it fixed everywhere, not just in the Reminders Settings. How would you suggest we approach it?

There are two more issues I noticed while reviewing it now:

- `BloggingRemindersStore` is not thread-safe, yet it gets called from a different thread (in theory, but unlikely in practice)
-  I think it may have another issue where the collection `blogs` is mutated during enumeration.

Reply:
> I'd say it's a good place to introduce SafeBlog or whatever we are going to call. How would you suggest we approach it?

I think that's a good idea in general: don't store nor pass NSManagedObject among functions. Depending on how many changes you'd like to make, replacing all places that use `Blog` with a `struct SafeBlog` would help other parts in blog reminder, and of course has a large blast radius. If you'd like to just resolve thread issue in this one single place, instead of using the `blog` argument, querying `Blog` instance using the `mainContext` inside the `DispatchQueue.main.async` seems to be sufficient enough.

Reply:
I think it's worth exploring the `SafeBlog` option, but maybe in a separate PR, because, as you said, it'll require a lot of code changes. I would like to avoid polluting the namespace with a new type unless it's clear there are at least a few good places where it can be used.

I followed the second suggestion: re-fetching the blog object on the main thread.